spec/message.abnf

```diff
@@ -0,0 +1,57 @@
+message = [s] *(declaration [s]) body [s]
+
+declaration = let s variable [s] "=" [s] "{" [s] expression [s] "}"
+body = pattern
+     / (selectors 1*([s] variant))
+
+pattern = "{" *(text / placeholder) "}"
+selectors = match 1*([s] selector)
+selector = "{" [s] expression [s] "}"
+variant = when 1*(s key) [s] pattern
+key = nmtoken / literal / "*"
+
+placeholder = "{" [s] (expression / markup / markup-end) [s] "}"
+
+expression = ((literal / variable) [s annotation])
+           / annotation
+annotation = function *(s option)
+option = name [s] "=" [s] (literal / nmtoken / variable)
+
+markup = markup-start *(s option)
+
+let = %x6C.65.74
+match = %x6D.61.74.63.68
+when = %x77.68.65.6E
+
+text = 1*(text-char / text-escape)
+text-char = %x0-5B         ; omit \
+          / %x5D-7A        ; omit {
+          / %x7C           ; omit }
+          / %x7E-D7FF      ; omit surrogates
+          / %xE000-10FFFF
+
+literal = "(" *(literal-char / literal-escape) ")"
+literal-char = %x0-27         ; omit ( and )
+             / %x2A-5B        ; omit \
+             / %x5D-D7FF      ; omit surrogates
+             / %xE000-10FFFF
+
+variable = "$" name
+function = ":" name
+markup-start = "+" name
+markup-end = "-" name
+
+name = name-start *name-char
+nmtoken = 1*name-char
+name-start = ALPHA / "_"
+           / %xC0-D6 / %xD8-F6 / %xF8-2FF
+           / %x370-37D / %x37F-1FFF / %x200C-200D
+           / %x2070-218F / %x2C00-2FEF / %x3001-D7FF
+           / %xF900-FDCF / %xFDF0-FFFD / %x10000-EFFFF
+name-char = name-start / DIGIT / "-" / "." / %xB7
+          / %x0300-036F / %x203F-2040
+
+text-escape = "\\" / "\{" / "\}"
+literal-escape = "\\" / "\(" / "\)"
```

Comment:
I recognize that this is in conflict with https://github.com/unicode-org/message-format-wg/pull/347#discussion_r1119976120 , but these strings are easy to misinterpret if the reader is not being careful. I personally like the approach used by [JSON](https://www.rfc-editor.org/rfc/rfc8259.html#section-7) of naming `\`:
```suggestion
text-escape    = backslash ( backslash / "{" / "}" )
literal-escape = backslash ( backslash / "(" / ")" )

backslash = %x5C ; U+005C REVERSE SOLIDUS "\"
```

Reply:
This is definitely easier to interpret. I did a couple head scratches when reading it in the diff here earlier, this would have been much less mental load to interpret.

Reply:
I thought that @eemeli's suggestion already improved the readability of these rules, but I'll agree that introducing `backslash` makes them even easier to understand. Thanks!